### PR TITLE
Cleaned up developer guide, added TOC

### PR DESCRIPTION
**Pull request by dblock**
Signed-off-by: dblock <dblock@amazon.com>

### Description

Cleaned up/edited developer guide. Might be easier to view [with w=1](https://github.com/opensearch-project/OpenSearch/pull/572/files?w=1).

- added a Visual Studio code section
- project layout was misplaced/split in parts
- added a git-secrets section, coming from #540 
- don't wrap lines inconsistently

### Check List

- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by anirudha — APPROVED** [review]
thanks! I foresee this is going to be expanded very soon; we can consider adding a ToC in the future.

**Comment by dblock** [comment]
> thanks! I foresee this is going to be expanded very soon; we can consider adding a ToC in the future.

Good idea, added in this PR.

**Comment by adnapibar** [comment]
start gradle check

**Comment by adnapibar** [comment]
start gradle check

**Comment by adnapibar** [comment]
start gradle check

File: DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -1,333 +1,279 @@
...
-```
+The canonical example of this is the ICU analysis plugin. It is important for folks who want the fairly language neutral ICU analyzer but the library to implement the analyzer is 11MB so we don't ship it with OpenSearch by default.
 
+Another example is the `discovery-gce` plugin. It is *vital* to folks running in [GCP](https://cloud.google.com/) but useless otherwise and it depends on a dozen extra jars.
 
-#### `qa`
+### `qa`
 
-Honestly this is kind of in flux and we're not 100% sure where we'll end up.  We welcome your throughts and help.
+Honestly this is kind of in flux and we're not 100% sure where we'll end up. We welcome your throughts and help.
```
Comment (adnapibar) [inline DEVELOPER_GUIDE.md:173]: This existing but I think we can fix it "throughts" -> "thoughts"

File: DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -1,333 +1,279 @@
...
+### Formatting Failures
 
-Unless your change is trivial, there will probably be discussion about the pull request and, if any changes are needed, we would love to work with you to get your pull request merged into OpenSearch.
+Sometimes Spotless will report a "misbehaving rule which can't make up its mind" and will recommend enabling the `paddedCell()` setting. If you enabled this settings and run the format check again, Spotless will write files to `$PROJECT/build/spotless-diagnose-java/` to aid diagnosis. It writes different copies of the formatted files, so that you can see how they
+differ and infer what is the problem.
+
+The `paddedCell()` option is disabled for normal operation in order to detect any misbehaviour. You can enabled the option from the command line by running Gradle with `-Dspotless.paddedcell`.
```
Comment (adnapibar) [inline DEVELOPER_GUIDE.md:231]: "can enabled" -> "can enable" ?

**Review by adnapibar — APPROVED** [review]
Thanks @dblock for this. LGTM! 👍 
Just a couple of typos.

File: DEVELOPER_GUIDE.md (hunk 1 above)
Comment (dblock) [reply DEVELOPER_GUIDE.md:173]: 🤦

**Comment by dblock** [comment]
start gradle check

**Comment by adnapibar** [comment]
start gradle check

**Pull request merged**